### DISCUSSION TITLE
Usefulness of an .app

### DISCUSSION THREAD

**Discussion by willcohen** [babashka/obb#9] [Ideas]
I was thinking this morning that an eventual option alongside the cli might be an `.app` as well -- it seems, at first glance, like the biggest outstanding difference between something like Hammerspoon and this is that `obb` still has no clear capability for running outside of an ad-hoc manner. If there were a canonical `obb.app` that could startup and listen for events, it could collect various ad-hoc additional scripts and respond to those events accordingly.

What I still haven't thought through in any detail is how that would work. Could it be as simple as, ultimately, a "fancy" .cljs script that itself uses obb to create a process that can load and unload scripts and perhaps even spit out a really simple GUI? If so, then I suppose the hardest part would be thinking through the eventual API for what can be listened to, and how scripts can be loaded and unloaded.

**Comment by willcohen** [comment]
It seems like JXA is able to listen to events as long as something like that is running: https://stackoverflow.com/questions/49501983/trigger-a-script-applescript-or-jxa-on-app-running. I'm not seeing anything socket-like built into JXA, so per @borkdude, the path might be to integrate something nrepl-y into an obb daemon, and optionally stick that into `obb.app,` which can load various smaller scripts as needed. There may still also be value in eventually having some syntactic sugar as additional namespaces in obb itself, to simplify common listener tasks.

**Comment by zane** [comment]
It is possible to compile JXA scripts themselves into `.app`s. Have a look at `man osacompile`. I've been considering adding a `--compile <script>` option to `obb` as a convenience, but I'm not sure whether that would be advisable as at that point one might be better off using the ClojureScript compiler and `osacompile` directly. I do think I would prefer that to having an official `obb.app`.

What do you think, @willcohen?

> …perhaps even spit out a really simple GUI?

Check out `examples/gui.cljs`!

**Reply by borkdude** [reply]
I guess one could osacompile obb itself + bash wrapper or so to make custom apps. I would not support this as an option in obb initially (because of the burden of having to support this while the idea is still too young) but would document it in doc/foobar.md or a wiki on this repo.

**Reply by zane** [reply]
I had imagined that `--compile <script>` would generate a program that runs `obb.impl.main/eval-string` on `<script>` and compile that directly to an `.app`, but (if I'm understanding you) what you're describing is an interesting alternative approach.

**Reply by borkdude** [reply]
Yeah, what you could also do is append some JS to the original obb which invokes the script. E.g. just append:
```
obb.main.main("script.cljs");
```
(probably have to relativize the script name to the .app executable)

Also have to use `^:export` on the main function in that case.

And it would then package the whole thing in an `.app`. This could just be a babashka or bash script or so probably that people can copy from the docs, in the initial stage of this feature.

**Comment by borkdude** [comment]
@willcohen See https://github.com/babashka/obb/blob/main/doc/application-bundle.md
Might not be totally what you were getting at, but possible related.
